File: CombatTracker/CombatTracker.Entities/Reference/Creatures/Charts/ConstitutionBonusChart.cs

using System.Collections.Generic;
using System.Linq;

namespace CombatTracker.Entities.Reference.Creatures.Charts
{
    public partial class ConstitutionBonusChart
    {
        public ConstitutionBonusChart()
        {
            this.ConstitutionBonusChartValues = new HashSet<ConstitutionBonusChartValues>();
            //this.Creatures = new HashSet<Creature>();
        }
    
        public int ID { get; set; }
        public string Code { get; set; }
        public int PerLevelDifference { get; set; }
        public int BonusExhaustion { get; set; }
    
        public virtual ICollection<ConstitutionBonusChartValues> ConstitutionBonusChartValues { get; set; }
        //public virtual ICollection<Creature> Creatures { get; set; }


        public override string ToString()
        {
            return Code;
        }

        //public static ConstitutionBonusChart LoadByName(RmssDbContext db, string name)
        //{
        //    return (from sc in db.ConstitutionBonusCharts where sc.Code == name select sc).FirstOrDefault();
        //}

        public int LookupStamina(int roll)
        {
            return (from cbc 
                        in ConstitutionBonusChartValues 
                    where cbc.Min <= roll && roll <= cbc.Max 
                    select cbc.Mod).FirstOrDefault();
        }
    }
    
}

File: CombatTracker-Application/CombatTrackerData/CreatureCharts/ConstitutionBonusChart.vb
Partial Class ConstitutionBonusChart
    Inherits IEntity(Of ConstitutionBonusChart)
    Public Overrides Function ToString() As String
        Return Code
    End Function

    Public Shared Function LoadByName(ByVal DB As RMSSDataDataContext, ByVal Name As String) As ConstitutionBonusChart
        Return (From SC In DB.ConstitutionBonusCharts Where SC.Code = Name Select SC).FirstOrDefault
    End Function

    Public Function LookupStamina(ByVal Roll As Integer) As Integer
        Return (From CBC In Me.ConstitutionBonusChartValues Where CBC.Min <= Roll AndAlso Roll <= CBC.Max Select CBC.Mod).FirstOrDefault
    End Function

    Private Sub OnValidate(ByVal action As System.Data.Linq.ChangeAction)
        If action <> Data.Linq.ChangeAction.Delete AndAlso Not IsValid Then
            Throw New ApplicationException("Rule violations prevent saving")
        End If
    End Sub

    Public Overrides Function GetRuleViolations() As System.Collections.Generic.List(Of RuleViolation)
        Dim L As New List(Of RuleViolation)
        'If String.IsNullOrEmpty(Name) Then
        '    L.Add(New RuleViolation("Name Required", "Name"))
        'End If
        Return L
    End Function
End Class
